streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -611,33 +545,38 @@ public final <KIn, VIn, VOut> void addProcessor(final String name,
 
     public final void addStateStore(final StoreBuilder<?> storeBuilder,
                                     final String... processorNames) {
+        addStateStore(new StoreBuilderWrapper<>(storeBuilder), false, processorNames);
+    }
+
+    public final void addStateStore(final StoreFactory<?> storeBuilder,
+                                    final String... processorNames) {
         addStateStore(storeBuilder, false, processorNames);
     }
 
-    public final void addStateStore(final StoreBuilder<?> storeBuilder,
+    public final void addStateStore(final StoreFactory<?> storeFactory,
                                     final boolean allowOverride,
                                     final String... processorNames) {
-        Objects.requireNonNull(storeBuilder, "storeBuilder can't be null");
-        final StateStoreFactory<?> stateFactory = stateFactories.get(storeBuilder.name());
-        if (!allowOverride && stateFactory != null && !stateFactory.builder.equals(storeBuilder)) {
-            throw new TopologyException("A different StateStore has already been added with the name " + storeBuilder.name());
+        Objects.requireNonNull(storeFactory, "stateStoreFactory can't be null");
+        final StoreFactory<?> stateFactory = stateFactories.get(storeFactory.name());
+        if (!allowOverride && stateFactory != null && !stateFactory.isCompatibleWith(storeFactory)) {
+            throw new TopologyException("A different StateStore has already been added with the name " + storeFactory.name());
         }
-        if (globalStateBuilders.containsKey(storeBuilder.name())) {
-            throw new TopologyException("A different GlobalStateStore has already been added with the name " + storeBuilder.name());
+        if (globalStateBuilders.containsKey(storeFactory.name())) {
+            throw new TopologyException("A different GlobalStateStore has already been added with the name " + storeFactory.name());
         }
 
-        stateFactories.put(storeBuilder.name(), new StateStoreFactory<>(storeBuilder));
+        stateFactories.put(storeFactory.name(), storeFactory);
 
         if (processorNames != null) {
             for (final String processorName : processorNames) {
                 Objects.requireNonNull(processorName, "processor name must not be null");
-                connectProcessorAndStateStore(processorName, storeBuilder.name());
+                connectProcessorAndStateStore(processorName, storeFactory.name());
             }
         }
         nodeGroups = null;
     }
 
-    public final <KIn, VIn> void addGlobalStore(final StoreBuilder<?> storeBuilder,
+    public final <KIn, VIn> void addGlobalStore(final StoreFactory<?> storeBuilder,
```

Comment:
nit: change the parameter name to `storeFactory`

Reply:
yup, there were lots of places I need to make this change but I didn't want to touch 2 million lines of code until the general strategy was looked at 😆

Reply:
very understandable -- just wanted to leave a comment in case you missed it/to help identify some of those two million lines 😭